_posts/2018-01-26-grpc-and-deadlines.md

```diff
@@ -0,0 +1,139 @@
+
+In general, when you don't set a deadline resources will be held for all in-flight requests, and all requests can potentially reach the maximum timeout. This puts the service at risk of running out of resources like memory and crashing, as well as maximizing the latency of the service.
+
+To avoid this, services should specify the longest default deadline they technically support, and clients should wait until the response is no longer useful to them. For the service this can be as simple as providing a comment in the .proto file. For the client this involves setting useful deadlines.
+
+There is no single answer to "What is a good deadline/timeout value?". Your service might be as simple as the Greeter in our quick start guides, in which case 100 ms would be fine. Your service might be as complex as a globally-distributed and strongly consistent database. The deadline for a client query will be different from how long they should wait for you to drop their table.
+
+So what do you need to consider to make an informed choice of deadline? Factors to take into account include the end to end latency of the whole system, which RPCs are serial, and which can be made in parallel. You should to be able to put numbers on it, even if it's a rough calculation. Engineers need to understand the service and then set a deliberate deadline for the RPCs between clients and servers.
+
+In gRPC, both the client and server make their own independent and local determination about whether the remote procedure call (RPC) was successful. This means their conclusions may not match! An RPC that finished successfully on the server side can fail on the client side. For example, the server can send the response, but the reply can arrive at the client after their deadline has expired. The client will already have terminated with the status error `DEADLINE_EXCEEDED`.
+
+## Setting a deadline
+
+As a client you should always set a deadline for how long you are willing to wait for a reply from the server. Here's an example using the greeting service from our [Quick Start Guides]({{ site.baseurl }}/docs/quickstart/):
+
+**C++**
+
+
+```
+  ClientContext context;
+  time_point deadline = std::chrono::system_clock::now() + 
+      std::chrono::milliseconds(100);
+  context.set_deadline(deadline);
+```
+
+
+**GO LANG**
+
+
+```
+    clientDeadline := time.Now().Add(time.Duration(*deadlineMs) * time.Millisecond)
+    ctx, cancel := context.WithDeadline(ctx, clientDeadline)
+```
+
+
+**JAVA**
+
+
+```
+response = blockingStub.withDeadlineAfter(deadlineMs, TimeUnit.MILLISECONDS).sayHello(request);
+```
+
```

Comment:
Should we add an example with Timeouts for completeness? Using python perhaps something like
```
response = stub.Foo(request,  timeout= 0.1)  # timeouts are specified in seconds
```
@nathanielmanistaatgoogle does this sound right

Reply:
I'm mostly ambivalent but leaning yes.